stl/inc/xutility

```diff
@@ -2426,30 +2429,73 @@ namespace ranges {
     _EXPORT_STD template <range _Rng>
     using range_rvalue_reference_t = iter_rvalue_reference_t<iterator_t<_Rng>>;
 
+#if _HAS_CXX23
+    _EXPORT_STD template <class _Rng>
+    concept constant_range = input_range<_Rng> && _Constant_iterator<iterator_t<_Rng>>;
+
+    template <range _Rng> // Per LWG-3768
+    _NODISCARD constexpr auto& _Possibly_const_range(_Rng& _Range) noexcept {

```

Comment:
There are test failures; one of them involves this line. (I haven't investigated the root cause.)

```
[...]
Command: "C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.34.31921\bin\HostX64\x64\cl.EXE" "D:\a\_work\1\s\llvm-project\libcxx\test\std\ranges\range.access\begin.pass.cpp" "-ID:\build\out\inc" "-ID:\a\_work\1\s\llvm-project\libcxx\test\support" "/nologo" "/Od" "/W4" "/w14061" "/w14242" "/w14265" "/w14582" "/w14583" "/w14587" "/w14588" "/w14749" "/w14841" "/w14842" "/w15038" "/w15214" "/w15215" "/w15216" "/w15217" "/w15262" "/sdl" "/WX" "/Zc:strictStrings" "/D_ENABLE_STL_INTERNAL_CHECK" "/bigobj" "/EHsc" "/MTd" "/std:c++latest" "/permissive-" "/FImsvc_stdlib_force_include.h" "/wd4643" "/D_STL_CALL_ABORT_INSTEAD_OF_INVALID_PARAMETER" "/D_USE_JOIN_VIEW_INPUT_RANGE" "/analyze:autolog-" "/Zc:preprocessor" "-FeD:\build\tests\libcxx\std\ranges\range.access\Output\begin.pass.cpp.dir\0\begin.pass.exe" "-link" "-LIBPATH:D:\build\out\lib\amd64" "-LIBPATH:C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.34.31921\lib\x64" "/MANIFEST:EMBED"
Exit Code: 2
Standard Output:
--
begin.pass.cpp
D:\a\_work\1\s\llvm-project\libcxx\test\std\ranges\range.access\begin.pass.cpp(33): error C2607: static assertion failed
D:\build\out\inc\xutility(2115): error C2338: static_assert failed: 'The range access customization point objects std::ranges::begin, std::ranges::end, std::ranges::rbegin, std::ranges::rend, and std::ranges::data do not accept arrays with incomplete element types.'
D:\build\out\inc\xutility(2130): note: see reference to function template instantiation 'std::_Choice_t<std::ranges::_Begin::_Cpo::_St> std::ranges::_Begin::_Cpo::_Choose<Incomplete(&)[10]>(void) noexcept' being compiled
D:\build\out\inc\xutility(2134): note: see reference to variable template 'const std::_Choice_t<std::ranges::_Begin::_Cpo::_St> std::ranges::_Begin::_Cpo::_Choice<Incomplete (&)[10]>' being compiled
D:\build\out\inc\xutility(2436): note: see reference to variable template 'bool range<Incomplete [10]>' being compiled
D:\build\out\inc\type_traits(1678): note: see reference to alias template instantiation 'std::_Decltype_invoke_nonzero<RangeCBeginT,Incomplete(&&)[10],>' being compiled
D:\a\_work\1\s\llvm-project\libcxx\test\std\ranges\range.access\begin.pass.cpp(43): note: see reference to variable template 'const bool is_invocable_v<std::ranges::_Cbegin_fn const ,Incomplete (&&)[10]>' being compiled
[...]
```

Reply:
I've investigated `std/containers/views/views.span/types.pass.cpp` failure - looks like C++23 makes it impossible to create some spans:

```c++
#include <span>
#include <string>

using namespace std;

// Fine in C++20; failure in C++23, because `decltype(s)::const_iterator`
// cannot be instantiated (`decltype(s)::iterator` is not
// `input_iterator`: https://godbolt.org/z/3vv9b8G3K).
int main() { [[maybe_unused]] span<const volatile string> s; }
```

CPOs' failures are a little bit more complicated...

Reply:
LWG-3768 is considered NAD, so there is no need to implement it.

Reply:
Reduced failing example:
```C++
#include <iterator>
struct S {}; // Notes: no constructor from volatile glvalues is implicitly declared
static_assert(std::input_iterator<const volatile S*>); // Fails: S is not constructible from const volatile S&
```

I think we need an LWG issue to fix `span` (maybe together with `input_iterator<volatile T*>` - it is arguably nonsense that `input_iterator<volatile int*>` is satistied but `input_iterator<volatile MyClass*>` is not).

Reply:
> D:\a\_work\1\s\llvm-project\libcxx\test\std\ranges\range.access\begin.pass.cpp(43): note: see reference to variable template 'const bool is_invocable_v<std::ranges::_Cbegin_fn const ,Incomplete (&&)[10]>' being compiled

This message is strange. `Incomplete (&&)[10]` should have been rejected by `_Should_range_access` in a SFINAE-friendly manner.

Reply:
Well, using `ranges::begin` with array that has elements of incomplete type is IFNDR ([
[range.access.begin]/2.2](https://eel.is/c++draft/range.access.begin#2.2)). So maybe this error is correct?

Similarly `ranges::range` concept cannot be used with such arrays: https://godbolt.org/z/Grfj1Ke4b

Reply:
Anyway, I'm going to skip those tests, since there are other failures. Like this one:

```c++
#include <ranges>
// Similar to "ranges/range.access/rend.pass.cpp:71"
struct Different {
  char *&begin();
  char *&end();
  short *&begin() const;
  short *&end() const;
} x;

// Ok in C++20, failure in C++23 -- we need to wait for libc++ to implement P2278R4
static_assert(std::same_as<decltype(std::ranges::cbegin(x)), short *>);
```

Reply:
Resolving this conversation as the tests are passing - please let me know if any additional action needs to be taken here. Thanks everyone for looking into these incredibly complex issues! :heart_eyes_cat: